File: Sources/ClaudeRelayClient/Views/AgentStatePill.swift
import SwiftUI
import ClaudeRelayKit

/// Pure color/word mapping for the agent-state pill. "Waiting" is the display
/// label for `.idle` (agent running, awaiting input).
public enum AgentStatePillModel {
    public static func word(_ s: AgentDetectedState) -> String {
        switch s {
        case .idle: return "Waiting"
        case .working: return "Working"
        case .blocked: return "Blocked"
        case .unknown: return "Unknown"
        }
    }

    public static func color(_ s: AgentDetectedState, agentId: String?, seen: Bool) -> Color {
        switch s {
        case .blocked: return .red
        case .working: return AgentColorPalette.color(for: agentId)
        case .idle: return seen ? .green : .yellow
        case .unknown: return .gray
        }
    }
}

/// Colored capsule + text word for a running agent's state.
public struct AgentStatePill: View {
    public let agentState: AgentDetectedState
    public let agentId: String?
    public let seen: Bool

    public init(agentState: AgentDetectedState, agentId: String?, seen: Bool) {
        self.agentState = agentState
        self.agentId = agentId
        self.seen = seen
    }

    public var body: some View {
        let c = AgentStatePillModel.color(agentState, agentId: agentId, seen: seen)
        Text(AgentStatePillModel.word(agentState))
            .font(.caption2)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(c.opacity(0.15))
            .foregroundStyle(c)
            .clipShape(Capsule())
    }
}

File: CodeRelayAndroid/feature-workspace/src/main/kotlin/relay/feature/workspace/ui/AgentStatePill.kt
package relay.feature.workspace.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import relay.protocol.AgentDetectedState
import relay.protocol.SessionState

/** Display word for the agent-state pill. IDLE is shown as "Waiting". */
fun agentStateWord(s: AgentDetectedState): String = when (s) {
    AgentDetectedState.IDLE -> "Waiting"
    AgentDetectedState.WORKING -> "Working"
    AgentDetectedState.BLOCKED -> "Blocked"
    AgentDetectedState.UNKNOWN -> "Unknown"
}

/** Pill color, parity with Swift AgentStatePillModel.color. */
fun agentStatePillColor(s: AgentDetectedState, agentId: String?, seen: Boolean): Color = when (s) {
    AgentDetectedState.BLOCKED -> QualityRed
    AgentDetectedState.WORKING -> agentColor(agentId)
    AgentDetectedState.IDLE -> if (seen) QualityGreen else IdleYellow
    AgentDetectedState.UNKNOWN -> UnknownGray
}

/** Left dot color for the session lifecycle/attachment state; null = not shown. */
fun sessionStatusDotColor(state: SessionState): Color? = when (state) {
    SessionState.ACTIVE_ATTACHED -> QualityGreen
    SessionState.ACTIVE_DETACHED, SessionState.CREATED,
    SessionState.STARTING, SessionState.RESUMING -> QualityYellow
    SessionState.EXITED, SessionState.FAILED,
    SessionState.TERMINATED, SessionState.EXPIRED -> null
}

@Composable
fun AgentStatePill(agentState: AgentDetectedState, agentId: String?, seen: Boolean) {
    val c = agentStatePillColor(agentState, agentId, seen)
    Text(
        text = agentStateWord(agentState),
        style = MaterialTheme.typography.labelSmall,
        color = c,
        modifier = Modifier
            .background(color = c.copy(alpha = 0.15f), shape = RoundedCornerShape(50))
            .padding(horizontal = 6.dp, vertical = 2.dp),
    )
}

@Composable
fun SessionStatusDot(state: SessionState, size: Dp = 8.dp) {
    val color = sessionStatusDotColor(state) ?: return
    Box(
        modifier = Modifier
            .size(size)
            .clip(CircleShape)
            .background(color),
    )
}
